### ISSUE TITLE
salt-minion segmentation fault

### ISSUE THREAD

**Issue by bahadunn** [saltstack/salt#31021]
```
open("/lib64/libkeyutils.so.1", O_RDONLY) = 6
read(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0`\n\0\0\0\0\0\0"..., 832) = 832
fstat(6, {st_mode=S_IFREG|0755, st_size=7176, ...}) = 0
mmap(NULL, 2102416, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 6, 0) = 0x7f3196513000
mprotect(0x7f3196515000, 2093056, PROT_NONE) = 0
mmap(0x7f3196714000, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 6, 0x1000) = 0x7f3196714000
close(6)                                = 0
open("/lib64/libresolv.so.2", O_RDONLY) = 6
read(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0203\0\0\0\0\0\0"..., 832) = 832
fstat(6, {st_mode=S_IFREG|0755, st_size=89880, ...}) = 0
mmap(NULL, 2181896, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 6, 0) = 0x7f31962fe000
mprotect(0x7f319630f000, 2097152, PROT_NONE) = 0
mmap(0x7f319650f000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 6, 0x11000) = 0x7f319650f000
mmap(0x7f3196511000, 6920, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f3196511000
close(6)                                = 0
open("/lib64/libselinux.so.1", O_RDONLY) = 6
read(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\340E\0\0\0\0\0\0"..., 832) = 832
fstat(6, {st_mode=S_IFREG|0755, st_size=92960, ...}) = 0
mmap(NULL, 2192800, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 6, 0) = 0x7f31960e6000
mprotect(0x7f31960fb000, 2097152, PROT_NONE) = 0
mmap(0x7f31962fb000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 6, 0x15000) = 0x7f31962fb000
mmap(0x7f31962fd000, 1440, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f31962fd000
close(6)                                = 0
open("/lib64/libsepol.so.1", O_RDONLY)  = 6
read(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0@=\0\0\0\0\0\0"..., 832) = 832
fstat(6, {st_mode=S_IFREG|0755, st_size=245232, ...}) = 0
mmap(NULL, 2383168, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 6, 0) = 0x7f3195ea0000
mprotect(0x7f3195edb000, 2097152, PROT_NONE) = 0
mmap(0x7f31960db000, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 6, 0x3b000) = 0x7f31960db000
mmap(0x7f31960dc000, 40256, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f31960dc000
close(6)                                = 0
mprotect(0x7f319650f000, 4096, PROT_READ) = 0
mprotect(0x7f3197465000, 12288, PROT_READ) = 0
access("/etc/selinux/", F_OK)           = 0
open("/etc/selinux/config", O_RDONLY)   = -1 ENOENT (No such file or directory)
open("/proc/mounts", O_RDONLY)          = 6
fstat(6, {st_mode=S_IFREG|0444, st_size=0, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f319dd09000
read(6, "rootfs / rootfs rw 0 0\n/dev/root"..., 1024) = 544
read(6, "", 1024)                       = 0
close(6)                                = 0
munmap(0x7f319dd09000, 4096)            = 0
munmap(0x7f319de02000, 33435)           = 0
close(5)                                = 0
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0x11} ---
+++ killed by SIGSEGV +++
Segmentation fault
```

**Comment by ssgward** [comment]
@bahadunn - what command are you running to get this error?

And what version of salt are you using? ``salt --versions-report``

**Comment by bahadunn** [comment]
running --version-report results in this:

[root@xen1 ~]# salt-minion --version-report
Segmentation fault
[root@xen1 ~]#

[root@xen1 ~]# rpm -qa | grep salt
salt-minion-2015.8.5-1.el5
salt-2015.8.5-1.el5
salt-repo-2015.8-2.el5
[root@xen1 ~]#

We have hosts with salt version 2015.5.0 that are fine.  The Segmentation
faults started with version 2015.5.3 and still happen even with 2015.8.5 as
you can see above.

Thanks,
Jason

**Comment by bahadunn** [comment]
Here is some more info that might be helpful.

[root@xen1 ~]# python2.6
Python 2.6.8 (unknown, Nov  7 2012, 14:47:45) 
[GCC 4.1.2 20080704 (Red Hat 4.1.2-52)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from salt.scripts import salt_minion
>>> from salt.utils import is_windows
Segmentation fault

**Comment by bahadunn** [comment]
More hopefully useful info:

Feb 10 12:59:46 xen1 kernel: [694500.664249] python2.6[26753]: segfault at 11 ip 00007f8f00b5f7d0 sp 00007fffe82b5c50 error 4 in libcrypto.so.1.0.1e[7f8f00a6d000+1a9000]
Feb 10 13:12:59 xen1 kernel: [695293.270926] python2.6[31105]: segfault at 11 ip 00007f713d0ae7d0 sp 00007fff516c4840 error 4 in libcrypto.so.1.0.1e[7f713cfbc000+1a9000]

**Comment by cachedout** [comment]
Could you do a `yum whatprovides libcrypto.so.1.0.1e` please?

**Comment by bahadunn** [comment]
[root@xen1 ~]# yum whatprovides libcrypto.so.1.0.1e
Loaded plugins: dellsysid, fastestmirror
Loading mirror speeds from cached hostfile
citrix                                                                                                               |  951 B     00:00     
dell-omsa-indep                                                                                                      | 1.9 kB     00:00     
dell-omsa-specific                                                                                                   | 1.9 kB     00:00     
salt-2015.8                                                                                                          |  951 B     00:00     
Warning: 3.0.x versions of yum would erroneously match against filenames.
 You can use "*/libcrypto.so.1.0.1e" and/or "*bin/libcrypto.so.1.0.1e" to get that behaviour
No Matches found
[root@xen1 ~]#

**Comment by cachedout** [comment]
Hrm. You might need to do a little massaging of that filename to figure out what yum is looking for. Basically, I'm trying to suss out which package thinks it owns that libcrypto library. This *sounds* like a conflict between one of the python crypto libs and python, but it's a little hard to say.

**Comment by bahadunn** [comment]
[root@xen1 ~]# yum whatprovides "*/libcrypto.so"
Loaded plugins: dellsysid, fastestmirror
Loading mirror speeds from cached hostfile
citrix/filelists                                                                                                     |  166 B     00:00     
dell-omsa-indep/filelists_db                                                                                         | 635 kB     00:00     
dell-omsa-specific/filelists_db                                                                                      | 4.1 kB     00:00     
salt-2015.8/filelists                                                                                                |  78 kB     00:00     
likewise-open-libs-5.4.0.60874-1xs77.48.x86_64 : Likewise Utility Libraries
Repo        : installed
Matched from:
Filename    : /opt/likewise/lib64/libcrypto.so

**Comment by cachedout** [comment]
Ah. I'm guessing that's the source of the problem here. Salt is calling out to pycrypto which is looking to the libcrypto shared object which isn't compatible with the version of python that's installed. You could probably verify this by pulling up a 2.6 python shell and doing an `import Crypto` and see if it segfaults.

**Comment by bahadunn** [comment]
[root@xen1 ~]# python26
Python 2.6.8 (unknown, Nov  7 2012, 14:47:45) 
[GCC 4.1.2 20080704 (Red Hat 4.1.2-52)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import Crypto
>>>

**Comment by cachedout** [comment]
@bahadunn I only got part of that. Did the interpreter exit or segfault or did it keep running?

**Comment by bahadunn** [comment]
it's doing something weird.  The interpreter did not segfault.  It kept running.

**Comment by bahadunn** [comment]
1:openssl-xs-libs-1.0.1e-42.xs15.x86_64 : A general purpose cryptography library with TLS
                                        : implementation
Repo        : installed
Matched from:
Filename    : /usr/lib64/libcrypto.so.10
Filename    : /usr/lib64/libcrypto.so.1.0.1e

**Comment by cachedout** [comment]
@bahadunn The problem might only be exercised by certain crypto functions so importing it *might* not be enough to reproduce the issue.. :-/

**Comment by bahadunn** [comment]
Just found this:

[root@xen1 /]# python2.6
Python 2.6.8 (unknown, Nov  7 2012, 14:47:45) 
[GCC 4.1.2 20080704 (Red Hat 4.1.2-52)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import hashlib
Segmentation fault

**Comment by cachedout** [comment]
Nice! So, clearly there's a conflict between the openssl install you have and your python install, since that's a part of the python stdlib. I'm a little concerned that this got introduced with your salt upgrade though. I assume this was the point where you switched to using `repo.saltstack.com`?

**Comment by bahadunn** [comment]
Well, this first started with version 2015.5.3 of salt-minion I believe.  We obtained that from a copr repo.  I found and installed the repo.saltstack.com repo in an effort to upgrade and see if that would fix it.

**Comment by cachedout** [comment]
Is this happening on every Xen box you have or is this just an artifact of an upgrade on this single machine?

**Comment by bahadunn** [comment]
This is happening on all new installs for us on xenserver version:

[root@xen1 ~]# cat /etc/redhat-release 
XenServer release 6.5.0-90233c (xenenterprise)
[root@xen1 ~]#

Xenserver running salt-minion version 2015.5.0 seems to be fine.  We had hacked the bootstrap script for saltstack to work on xenserver which was fine until recently when the copr repo seemed to have moved.  Ever since then when salt minion upgraded to 2015.5.3 in those repos it's been segfaulting.  We ran into the issue when installing on new deployments of xenserver though.  I'm sure if we upgraded to 2015.5.3 on in production xenservers we'd see the segfault.  We've not tried this to confirm though.

**Comment by bahadunn** [comment]
I think I might have an idea as to the problem. My hacked up install.sh script has an override:
    if [ "${DISTRO_NAME}" = "XenServer" ] ; then
        DISTRO_NAME="CentOS"
        DISTRO_VERSION="5"
    fi

CentOS5 maxes out at openssl 0.9.something
But Citrix, with their out of band patches, they appear to have manually compiled that new openssl-xs
which is 1.0.1e
So if we're installing a python26 version from salt, that's compiled for stock centos5, it's definitely going to be expecting openssl 0.9.8
That bootstrap script has tons of "yum -y" commands in it, so I'm fearful of hacking my hacked script to specify         DISTRO_VERSION="6" (which would have python compiled against openssl 1.0.1e)
not sure which other packages it might take the liberty of upgrading for us, which may or may not kill xenserver

**Comment by cachedout** [comment]
Yup, that sounds like exactly the problem.

**Comment by trenta** [comment]
We are seeing this same issue in an upgrade to our XenServer environment from 6.2.0 to 6.5.0.  Thanks for this info and I'm looking forward to a solution.

**Comment by Criggie** [comment]
Concur with @trenta   I've got four xenserver 6.2 hosts which were upgraded to 6.5 via a rolling pool upgrade rather than a fresh install.

Freshly-installed xs65 hosts have openssl-0.9.8e-36.0.1.el5_11
whereas the rolling upgraded ones have  openssl-xs-1.0.1e-51.xs17 and no openssl rpm.

```
[root@badhost ~]# salt-minion --versions-report
Segmentation fault
[root@badhost ~]# rpm -q salt-minion
salt-minion-2015.8.8-2.el5
```
Anytime salt-minion segfaults, the errorcode is 139

Failing host has these OpenSSL RPMs
```
[root@badhost ~]# rpm -qa | grep openssl
openssl-perl-0.9.8e-27.el5_10.4
openssl-xs-libs-1.0.1e-51.xs17
openssl-wrapper-0.1-59
openssl-xs-1.0.1e-51.xs17

```
A working XS65 host has these OpenSSL RPMs
[root@goodhost ~]# rpm -qa | grep openssl
openssl-perl-0.9.8e-36.0.1.el5_11
openssl-0.9.8e-36.0.1.el5_11
openssl-0.9.8e-36.0.1.el5_11


UPDATE:  openssl-xs and openssl-0.9.8e  do co-exist on some hosts, and there salt-minion works as expected.  However installing openssl comes up with this mess:

```
yum  --enablerepo base install openssl --skip-broken
...
Packages skipped because of dependency problems:
    device-mapper-1.02.67-2.el5.i386 from base
    e2fsprogs-1.39-37.el5.x86_64 from base
    e2fsprogs-libs-1.39-37.el5.i386 from base
    e2fsprogs-libs-1.39-37.el5.x86_64 from base
    glibc-2.5-123.i686 from base
    keyutils-libs-1.2-1.el5.i386 from base
    krb5-libs-1.6.1-78.el5.i386 from base
    krb5-libs-1.6.1-78.el5.x86_64 from base
    libselinux-1.33.4-5.7.el5.i386 from base
    libsepol-1.15.2-3.el5.i386 from base
    openssl-0.9.8e-27.el5_10.4.i686 from base
    openssl-0.9.8e-27.el5_10.4.x86_64 from base
    zlib-1.2.3-7.el5.i386 from base
``` 
And frankly that's getting scary.

**Comment by Criggie** [comment]
Well I'm not proud of this fix, but it got me going again and solved the problem

1) Do **not** uninstall openssl-xs even through nothing depends on it.  If you do , ssh, scp, wget, curl all break.

2) Visit    http://vault.centos.org/5.10/os/i386/CentOS/
Note that its 5.**10** not 5.11 or 5 (which is a symlink to the latest 5.11 anyway)
None of the "normal" repos will hold 5.10 RPMs any more.

Download these RPMs and these versions:
device-mapper-1.02.67-2.el5.i386.rpm        libselinux-1.33.4-5.7.el5.i386.rpm
e2fsprogs-libs-1.39-36.el5_9.i386.rpm         libsepol-1.15.2-3.el5.i386.rpm
glibc-2.5-123.el5_11.1.i686.rpm              openssl-0.9.8e-36.0.1.el5_11.i686.rpm
keyutils-libs-1.2-1.el5.i386.rpm            openssl-0.9.8e-36.0.1.el5_11.x86_64.rpm
krb5-libs-1.6.1-80.el5_11.i386.rpm           zlib-1.2.3-7.el5.i386.rpm

Some machines also demanded this RPM as well, some didn't.
glibc-common-2.5-123.el5_11.1.i386.rpm

3) Get all these RPMs to your xenserver(s)   I cheated and used my ISO mount at
/var/run/sr-mount/ff89dde0-2ad0-9b22-a80d-8f9f6ed9dc30/

4)   run this
rpm -i *rpm --force
to install all those RPMs.    The force is required because of collisions between openssl-xs  and openssl.

And then `salt-call state,highstate worked.` as did 

```
[root@c5220-3-3 rpms]# python26
Python 2.6.8 (unknown, Mar 15 2016, 21:23:59)
[GCC 4.1.2 20080704 (Red Hat 4.1.2-55)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import hashlib
>>>
```
See?  No segfault!

**Comment by censored** [comment]
I also ran into this issue, Citrix XenServer 6.5.
I built out my Xenserver using the latest image provided by Citrix. 
As part of my automated install, Salt is set up and configured and runs successfully.
I then applied the  patches to Xenserver 6.5 (the list is huge, but basically Service Pack 1 and everything after that). 
At this point, python26 import hashlib started the segfault, and Salt no longer functions.

Using the article above, I went to  http://vault.centos.org/5.10/os/i386/CentOS/ and obtained openssl-0.9.8e-26.el5_9.1.x86_64.rpm (I could not find 0.9.8e-36.0.1.el5_11.x86_64.rpm) and did a force install.

So far, so good - Salt is not segfaulting - but I can't say something else isn't going to blow up.
